ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,403 @@
+#include "shared-bindings/pulseio/PWMOut.h"
+#include "supervisor/shared/translate.h"
+
+#include "shared-bindings/microcontroller/__init__.h"
+#include "stm32f4xx_hal.h"
+#include "common-hal/microcontroller/Pin.h"
+
+#define PWM_MAX_FREQ 6000000
+#define ALL_CLOCKS 0xFFFF
+
+STATIC uint8_t reserved_tim[TIM_BANK_ARRAY_LEN];
+STATIC uint32_t tim_frequencies[TIM_BANK_ARRAY_LEN];
+STATIC bool never_reset_tim[TIM_BANK_ARRAY_LEN];
+
+STATIC void tim_clock_enable(uint16_t mask);
+STATIC void tim_clock_disable(uint16_t mask);
+
+// Get the frequency (in Hz)
+static uint32_t timer_get_source_freq(uint32_t tim_id) {
+    uint32_t source, clk_div;
+    if (tim_id == 1 || (8 <= tim_id && tim_id <= 11)) {
+        // TIM{1,8,9,10,11} are on APB2
+        source = HAL_RCC_GetPCLK2Freq();
+        clk_div = RCC->CFGR & RCC_CFGR_PPRE2;
+    } else {
+        // TIM{2,3,4,5,6,7,12,13,14} are on APB1
+        source = HAL_RCC_GetPCLK1Freq();
+        clk_div = RCC->CFGR & RCC_CFGR_PPRE1;
+    }
+    if (clk_div != 0) {
+        // APB prescaler for this timer is > 1
+        source *= 2;
+    }
+    return source;
+}
+
+void pwmout_reset(void) {
+    uint16_t never_reset_mask = 0x00;
+    for(int i=0;i<TIM_BANK_ARRAY_LEN;i++) {
+        if(!never_reset_tim[i]) {
+            reserved_tim[i] = 0x00;
+            tim_frequencies[i] = 0x00;
+        } else {
+            never_reset_mask |= 1<<i;
+        }
+    }
+    tim_clock_disable(ALL_CLOCKS & ~(never_reset_mask));
+}
+
+void common_hal_pulseio_pwmout_never_reset(pulseio_pwmout_obj_t *self) {
+    for(size_t i = 0 ; i < TIM_BANK_ARRAY_LEN; i++) {
+        if (mcu_tim_banks[i] == self->handle.Instance) {
+            never_reset_tim[i] = true;
+            never_reset_pin_number(self->tim->pin->port, self->tim->pin->number);
+            break;
+        }
+    }
+}
+
+void common_hal_pulseio_pwmout_reset_ok(pulseio_pwmout_obj_t *self) {
+    //TODO: doesn't this need an equivalent release pin in microcontroller.c? 
+    //I don't see that implemented in any port. 
+    for(size_t i = 0 ; i < TIM_BANK_ARRAY_LEN; i++) {
+        if (mcu_tim_banks[i] == self->handle.Instance) {
+            never_reset_tim[i] = false;
+            break;
+        }
+    }
+}
+
+pwmout_result_t common_hal_pulseio_pwmout_construct(pulseio_pwmout_obj_t* self,
+                                                    const mcu_pin_obj_t* pin,
+                                                    uint16_t duty,
+                                                    uint32_t frequency,
+                                                    bool variable_frequency) {
+    if (frequency == 0 || frequency > 6000000) {
+        mp_raise_ValueError(translate("Invalid frequency supplied"));
+        //return PWMOUT_INVALID_FREQUENCY;
+    }
+
+    TIM_TypeDef * TIMx;
+    //TODO: style/safety question, use this instead?
+    //uint8_t tim_num_pins = TIM_PIN_ARRAY_LEN;
+    uint8_t tim_num = sizeof(mcu_tim_pin_list)/sizeof(*mcu_tim_pin_list);
```

Comment:
EIther is fine, because it's fixed by the hardware. But if you can think of a future case where `mcu_tim_pin_list` is not all the timers, then the current code makes more sense.